### DISCUSSION TITLE
Smartwings WM25L-Z - Firmware and Device

### DISCUSSION THREAD

**Discussion by tube0013** [zigpy/zigpy#1184] [General]
@mattwestb Moving the discussion from https://github.com/zigpy/zha-device-handlers/pull/2220 over here where it is more appropriate.

> From one CLI with commander for Billy its
>
>`.\commander.exe readmem --range 0xfe00000:0xfe00800 -d EFR32MG1P233F256GM48 -o E2022UD.s37`
>You only need changing the device type and output format if you like have it in Bin, hex and so on (bin is easy looking in with >hex editor but have no CRC check) by changing the out file extension.
>The range is the same for MG1 and 2 devices for user data so it is working OK and i have dumping the IKEA MG 21 modules >with it :-))
>
>Segger Jlink Flasher is also working OK but its depends what you normally using.

I didn't know the memory addresses off hand last night so just used the `--region @mainflash` command. I think I can add in `@userdata` as well to get the user data pages. I'll hook it up to the debugger again tonight. shot me an email at store @ tubeszb.com and I'll send you the dump I don't think I should post here.

I have a si labs developer kit with the built in j-link. so have been using that. also use if for flashing new modules. 

here's the OTA next image query from the devices:

`[zigpy.zcl] [0xE22A:1:0x0019] OTA query_next_image handler for 'Smartwings WM25/L-Z': field_control=FieldControl.0, manufacturer_id=4098, image_type=0, current_file_version=2, hardware_version=None, model='WM25/L-Z'`

**Comment by MattWestb** [comment]
The region variable shall working but i was not getting it working then we was cooking one user data for IKEA Symfonisk 2 that is debug locked so we cant reading the flash but using Styrbars main flash (for the boot loader) and flashing the sniffed OTA file the device is working and looks being 100% genuine IKEA on one other MG210L module also for Diriger hub:-))

I sending you one store mail later then im interesting how is being done and it can being its made of one tuya factory firmware and  then patching it without tuya promise for going around paying the tuya cloud fees.

**Reply by MattWestb** [reply]
I was sniffing my TS004F then doing OTA update request and is was replaying MF 0x1002 = manufacturer_id=4098 so its 110% tuya as our **P** have saying :-)))

**Comment by tube0013** [comment]
here's what is viewable from the bin in hex editor for the region you gave me above:

![image](https://user-images.githubusercontent.com/11084412/226731375-25afd453-dfe2-464c-b5c6-ef83ce87104d.png)

**Reply by tube0013** [reply]
I start to interpret that as there is a uart connection with the motor with baudrate 9600 and tx on 6 and rx on 7

**Reply by MattWestb** [reply]
Im agree with you and some protocol version they is using and some CRC checksum used in the serial protocol.
I think in the end is more hard/software setting for the Zigbee module but they have not putting the manufacture name there what i can see but perhaps the TX power setting and some protocol version.
Did you finding the manufacture name and model in the main flash as strings ?

**Reply by tube0013** [reply]
I just sent you the userdata bin files from 2 modules, 1 new and one original. plus the s37 of the main flash - this is the one Pud saw the TZ.... in.

**Reply by MattWestb** [reply]
I have getting them and i was wrong i 2 things:
- 1 MG2X is having smaller userdata but it working OK dumping it then the last part is only zeros.
- 2 The data at the end is not hardware setting its the tuya keys token for flashing the device and is used for connecting to tuya cloud and billing the manufacture for the all request to it.

Was converting the main flash to bin file with `.\commander.exe convert .\fw.s37 -o fw.bin` and looks interesting at 0x2ebdb:
![SWings](https://user-images.githubusercontent.com/49618193/226737456-7b172590-e1fa-4fa8-b48a-16c4419af3d5.PNG)
So its one 110% tuya device with cloud functionality but the manufacture have paying extra for getting there name in the device.

And i was looking little more for the Puddly find and its in the end of the main flash after the real application with some other data so its _TZE200 is not in the firmware code its in the NVM but very likely written by the main application but it can being done in the factory.

But if looking at the en of the application is some interesting strings that you have seen in the userdata:
![SWingsA](https://user-images.githubusercontent.com/49618193/226740185-17f76d57-f3b8-446b-9976-56db5f361d24.PNG)
If reading it right is the variable you was finding in the user data and its also saying that the number is one firmware version and also GPO for being waked up and waking up the MCU so the blind is one tuya MCU device !!!

Perhaps we can finding the DPs in the firmware ?

**Reply by tube0013** [reply]
The device doesn't present with any of the tuya non-sense, and for the most part spec compliant - quirk just inverts commands and doubles battery. it communicates to MCU only over 2 wires. as the other 2 are ground and 3.3v.  the motor/mcu device can trigger zigbee joining by both a physical button on the motor (hold for 6 seconds) or the 433.92 mhz remote. beyond that the zigbee module is just programmed to do up/down and lift percent along with reporting lift percent. it doesn't seem to do anything else. 

If I was competent in writing code I'd get a scope and put it on those 2 lines, and figure out what was going on, then write new fw that included full group and scene support 😂
